### ISSUE TITLE
Invoke-NAVContainerCodeunit - Windows security identifier is not supported in online environments.

### ISSUE THREAD

**Issue by vvkjndl** [microsoft/navcontainerhelper#2534]
**Describe the issue**
On executing the command Invoke-NAVContainerCodeunit we get the error `Windows security identifier is not supported in online environments.`

This error can't be reproduced in BC20.1 or below, and only started to happen since we started testing BC20.2.

**Scripts used to create container and cause the issue**
```
New-BcContainer -accept_eula -accept_outdated -updateHosts -dns '8.8.8.8' `
    -artifactUrl $artifactUrl `
    -containerName $ContainerName `
    -multitenant `
    -enableTaskScheduler:$False `
    -auth UserPassword -Credential $credential `
    -memoryLimit $Memory `
    -isolation $isolation `
    -includeAL `
    -assignPremiumPlan `
    -alwaysPull `
    -licenseFile $flfFilename `
    -imageName $myimagename `
    -useSSL:$UseSSL `
    -installCertificateOnHost:$UseSSL `
    -runSandboxAsOnPrem:$runSandboxAsOnPrem `
    -additionalParameters @("--storage-opt size=$StorageLimit") `
```

**Full output of scripts**
*Container Build Logs*
```

Found version 20.2
Get-BcContainerNavVersion Telemetry Correlation Id: a54ee23e-0752-429d-9a90-e80e507cc39c
Using License:  C:\ProgramData\BcContainerHelper\INECTA\fin.flf
Getting Artifacts for : Version: 20.2.0.0 Country: us Type: sandbox
Using BCArtifacts:  https://bcartifacts.azureedge.net/sandbox/20.2.41144.41959/us
Removing old desktop shortcuts for tenant url(s)...
BcContainerHelper is version 3.0.10
BcContainerHelper is running as administrator
Hyper-V is Enabled
UsePsSession is True
Host is Microsoft Windows 10 Enterprise - 21H2
Docker Client Version is 20.10.14
Docker Server Version is 20.10.14
Removing Desktop shortcuts
Fetching all docker images
Fetching all docker volumes
ArtifactUrl and ImageName specified
INFO: Windows 10 21H1/21H2 images are not yet available, using 2004 as these are found to work better than 20H2 on 21H1/21H2
WARNING: Unable to find matching generic image for your host OS. Using mcr.microsoft.com/businesscentral:10.0.19041.1415
Image sandbox20.2.0.0us:sandbox-20.2.41144.41959-us-mt already exists
Using image sandbox20.2.0.0us:sandbox-20.2.41144.41959-us-mt
Creating Container vjtest1
Style: sandbox
Multitenant: Yes
Version: 20.2.41144.41959
Platform: 20.0.41086.41925
Generic Tag: 1.0.2.7
Container OS Version: 10.0.19041.1415 (2004)
Host OS Version: 10.0.19044.1706 (21H2)
Using hyperv isolation
Using locale en-US
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file C:\ProgramData\BcContainerHelper\INECTA\fin.flf
Additional Parameters:
--storage-opt size=20G
--env customNavSettings=EnableTaskScheduler=False
Files in C:\ProgramData\BcContainerHelper\Extensions\vjtest1\my:
- AdditionalOutput.ps1
- license.flf
- MainLoop.ps1
- SetupNavUsers.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container vjtest1 from image sandbox20.2.0.0us:sandbox-20.2.41144.41959-us-mt
d189b26745c83951631a248851730ce0179eb474e3960bf3f60ba8dcc25ee095
Waiting for container vjtest1 to be ready
Initializing...
Setting host.docker.internal to 10.6.0.4 in container hosts file (copy from host hosts file)
Setting gateway.docker.internal to 10.6.0.4 in container hosts file (copy from host hosts file)
Setting kubernetes.docker.internal to 127.0.0.1 in container hosts file (copy from host hosts file)
Setting host.containerhelper.internal to 172.25.16.1 in container hosts file
Starting Container
Hostname is vjtest1
PublicDnsName is vjtest1
Using NavUserPassword Authentication
Starting Local SQL Server
Starting Internet Information Server
Creating Self Signed Certificate
Self Signed Certificate Thumbprint 0CA5D6F48D935532BB37034FCF997C6A69001598
DNS identity vjtest1
Modifying Service Tier Config File with Instance Specific Settings
Modifying Service Tier Config File with settings from environment variable
Setting EnableTaskScheduler to False
Starting Service Tier
Registering event sources
Creating DotNetCore Web Server Instance
Using application pool name: BC
Using default container name: NavWebApplicationContainer
Copy files to WWW root C:\inetpub\wwwroot\BC
Create the application pool BC
Create website: NavWebApplicationContainer without SSL
Update configuration: navsettings.json
Done Configuring Web Client
Enabling Financials User Experience
Using license file 'c:\run\my\license.flf'
Import License
Dismounting Tenant
Mounting Tenant
Mounting Database for default on server localhost\SQLEXPRESS with AllowAppDatabaseWrite = False
Sync'ing Tenant
Tenant is Operational
Creating http download site
Setting SA Password and enabling SA
Creating Admin as SQL User and add to sysadmin
Creating SUPER user
Assign Premium plan for ADMIN
Container IP Address: 172.25.26.231
Container Hostname  : vjtest1
Container Dns Name  : vjtest1
Web Client          : http://vjtest1/BC/?tenant=default
Dev. Server         : http://vjtest1
Dev. ServerInstance : BC
Dev. Server Tenant  : default
Setting vjtest1 to 172.25.26.231 in host hosts file
Setting vjtest1-default to 172.25.26.231 in host hosts file
Setting vjtest1-default to 172.25.26.231 in container hosts file

Files:
http://vjtest1:8080/ALLanguage.vsix

WARNING: You are running a container which is 96 days old.
Microsoft recommends that you always run the latest version of our containers.

Container Total Physical Memory is 12.5Gb
Container Free Physical Memory is 10.6Gb

Initialization took 55 seconds
Ready for connections!
Reading CustomSettings.config from vjtest1
Creating Desktop Shortcuts for vjtest1
Creating .net Assembly Reference Folder for VS Code
Copying DLLs from C:\Windows\assembly to assemblyProbingPath
Copying DLLs from C:\Windows\Microsoft.NET\assembly to assemblyProbingPath
Copying DLLs from C:\Program Files\Microsoft Dynamics NAV\200\Service to assemblyProbingPath
Copying DLLs from C:\Program Files (x86)\Microsoft Dynamics NAV\200\RoleTailored Client to assemblyProbingPath
Copying DLLs from C:\Test Assemblies\Mock Assemblies to assemblyProbingPath
Copying DLLs from C:\Program Files (x86)\Open XML SDK to assemblyProbingPath
Container vjtest1 successfully created

Use:
Get-BcContainerEventLog -containerName vjtest1 to retrieve a snapshot of the event log from the container
Get-BcContainerDebugInfo -containerName vjtest1 to get debug information about the container
Enter-BcContainer -containerName vjtest1 to open a PowerShell prompt inside the container
Remove-BcContainer -containerName vjtest1 to remove the container again
docker logs vjtest1 to retrieve information about URL's again
Container Creation processing time :  00:01:24.6627359
Updating server configuration...
Installing font 'c:\Windows\Fonts\segoeui.ttf' succeeded
Skipping font 'c:\Windows\Fonts\segoeui.ttf' as it is already installed
```
*Invoke-NAVContainerCodeunit Logs*
```
Windows security identifier is not supported in online environments.

Exception Script Stack Trace:
at <ScriptBlock>, <No file>: line 37

PowerShell Call Stack:
at Invoke-ScriptInBcContainer, C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.10\ContainerHandling\Invoke-ScriptInNavContainer.ps1: line 58
at Invoke-NavContainerCodeunit, C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.10\ObjectHandling\Invoke-NavContainerCodeunit.ps1: line 49
at global:Install-SetupDataPack, C:\Users\azure-admin\Desktop\SCRIPTS\SIMPMgmt.ps1: line 3946
at <ScriptBlock>, <No file>: line 1

Container Free Physical Memory: 7.9Gb

Services in container vjtest1:
- MicrosoftDynamicsNavServer$BC is Running
- MSSQL$SQLEXPRESS is Running


Relevant event log from container vjtest1:
- 20220615 05:12:54 - MicrosoftDynamicsNavServer$BC

  Server instance: BC
  Category: Security
  ClientSessionId: fa1b38f9-fae6-465f-be37-53defbd47647
  ClientActivityId: 887d9137-86b6-4764-91b0-acff1d856ff8
  ServerSessionUniqueId: 6bcddb6a-bcf1-4278-81fa-c112e35268a1
  ServerActivityId: adbe6af3-6907-49f8-b714-ea6ef3192a1c
  EventTime: 06/15/2022 09:12:54
  Message (FaultException`1): RenameUser(String, String, String, NavUserState, DateTime, String, String, Boolean, LicenseType, Boolean, String, String, Boolean, DateTime, String, String, String, Guid, String, String, String, Nullable`1) FAILURE.
  RootException: FaultException`1
  ExceptionStackTrace:
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.ErrorMappingCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.PartnerTelemetryErrorCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase m, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.PartnerTelemetryPermissionErrorCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.PushPopCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.TelemetryCombinator(Category telemetryCategory, Verbosity telemetryVerbosity, ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
  CallerStackTrace:
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.TelemetryCombinator(Category telemetryCategory, Verbosity telemetryVerbosity, ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.PartnerTelemetryErrorCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase m, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.PartnerTelemetryPermissionErrorCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.PushPopCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.TelemetryCombinator(Category telemetryCategory, Verbosity telemetryVerbosity, ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass6_0.<Create>b__0(ServiceOperation operation, NSServiceBase instance, MethodBase syncMethodLocal, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationTracer.TraceScopeCombinator(Category telemetryCategory, ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass14_0.<PerformanceCounterCombinator>b__0()
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.PerformanceCounterCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.InitClientTelemetryIdsCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.TlsClearCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass32_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
     at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.Invoke(Object instance, Object[] inputs, Object[]& outputs)
     at System.ServiceModel.Dispatcher.DispatchOperationRuntime.InvokeBegin(MessageRpc& rpc)
     at System.ServiceModel.Dispatcher.ImmutableDispatchRuntime.ProcessMessage5(MessageRpc& rpc)
     at System.ServiceModel.Dispatcher.ImmutableDispatchRuntime.ProcessMessage11(MessageRpc& rpc)
     at System.ServiceModel.Dispatcher.MessageRpc.Process(Boolean isOperationContextSet)
     at System.ServiceModel.Dispatcher.ChannelHandler.DispatchAndReleasePump(RequestContext request, Boolean cleanThread, OperationContext currentOperationContext)
     at System.ServiceModel.Dispatcher.ChannelHandler.HandleRequest(RequestContext request, OperationContext currentOperationContext)
     at System.ServiceModel.Dispatcher.ChannelHandler.AsyncMessagePump(IAsyncResult result)
     at System.ServiceModel.Dispatcher.ChannelHandler.OnAsyncReceiveComplete(IAsyncResult result)
     at System.Runtime.Fx.AsyncThunk.UnhandledExceptionFrame(IAsyncResult result)
     at System.Runtime.AsyncResult.Complete(Boolean completedSynchronously)
     at System.ServiceModel.Channels.TransportDuplexSessionChannel.TryReceiveAsyncResult.OnReceive(IAsyncResult result)
     at System.Runtime.Fx.AsyncThunk.UnhandledExceptionFrame(IAsyncResult result)
     at System.Runtime.AsyncResult.Complete(Boolean completedSynchronously)
     at System.ServiceModel.Channels.SynchronizedMessageSource.ReceiveAsyncResult.OnReceiveComplete(Object state)
     at System.ServiceModel.Channels.SessionConnectionReader.OnAsyncReadComplete(Object state)
     at System.Runtime.Fx.AsyncThunk.UnhandledExceptionFrame(IAsyncResult result)
     at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
     at System.Net.LazyAsyncResult.ProtectedInvokeCallback(Object result, IntPtr userToken)
     at System.Net.Security.NegotiateStream.ProcessFrameBody(Int32 readBytes, Byte[] buffer, Int32 offset, Int32 count, AsyncProtocolRequest asyncRequest)
     at System.Net.Security.NegotiateStream.ReadCallback(AsyncProtocolRequest asyncRequest)
     at System.Net.AsyncProtocolRequest.CompleteRequest(Int32 result)
     at System.Net.FixedSizeReader.CheckCompletionBeforeNextRead(Int32 bytes)
     at System.Net.FixedSizeReader.ReadCallback(IAsyncResult transportResult)
     at System.Runtime.AsyncResult.Complete(Boolean completedSynchronously)
     at System.ServiceModel.Channels.ConnectionStream.IOAsyncResult.OnAsyncIOComplete(Object state)
     at System.Net.Sockets.SocketAsyncEventArgs.OnCompleted(SocketAsyncEventArgs e)
     at System.Net.Sockets.SocketAsyncEventArgs.FinishOperationSuccess(SocketError socketError, Int32 bytesTransferred, SocketFlags flags)
     at System.Net.Sockets.SocketAsyncEventArgs.CompletionPortCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped)
     at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pOVERLAP)


  ProcessId: 2592
  Tag: 00000ZX
  ThreadId: 76
  CounterInformation:
  CustomParameters: {
  }
  GatewayCorrelationId: 131e5674-ee9f-46a8-8569-60b8cf61d150

Invoke-NAVContainerCodeunit Telemetry Correlation Id: 6e8828bd-415f-47ef-99ed-0066b894239f
Windows security identifier is not supported in online environments.
At C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.10\ContainerHandling\Invoke-ScriptInNavContainer.ps1:99 char:13
+             throw $errorMessage
+             ~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Windows securit...e environments.:String) [], RuntimeException
    + FullyQualifiedErrorId : Windows security identifier is not supported in online environments.
```

**Screenshots**
N/A

**Additional context**
- does it happen all the time?
Yes, since BC20.2 release.
- did it use to work?
Works on BC20.1 and before releases.

**Comment by freddydk** [comment]
-runsandboxAsOnPrem must be false.

If that parameter is true, the output should say:
![image](https://user-images.githubusercontent.com/10775043/173920002-079e2036-33de-4732-8f3a-458c799f0633.png)

Please check that you don't have a spelling mistake or like.
Sandbox containers are running like saas and thus the error.
If that is to be fixed, then you would need to file a bug on Business Central, not something we can fix in BcContainerHelper.

**Comment by vvkjndl** [comment]
We are using sandbox images to build container.

`https://bcartifacts.azureedge.net/sandbox/20.2.41144.41959/us`

I have verified the parameter `-runsandboxAsOnPrem`, even with value set as `$True` the container style doesn't change to `onprem`.

**Comment by freddydk** [comment]
The code in New-BcContainer looks like this:

```
        $bcstyle = "onprem"
        if (!$runSandboxAsOnPrem -and ($appManifest.PSObject.Properties.name -eq "isBcSandbox")) {
            if ($appManifest.isBcSandbox) {
                $bcstyle = "sandbox"
                if (!($PSBoundParameters.ContainsKey('multitenant')) -and !$skipDatabase) {
                    $multitenant = $bcContainerHelperConfig.sandboxContainersAreMultitenantByDefault
                }
            }
        }
```

I cannot see how bcstyle is set to sandbox if runSandboxAsOnPrem is true.
Your container also becomes multi-tenant.

Please check for spelling mistakes or other syntax errors in the script - maybe the script doesn't include the line?
You could post a screenshot of the script with coloring?

**Comment by freddydk** [comment]
I have created a bug on the Business Central platform (internally) to make Invoke-NavCodeunit work with sandbox containers.

**Comment by vvkjndl** [comment]
I added a testing line in this codeblock and it is indeed changing the bcstyle back to sandbox:

```
    if ($navVersion -eq "") {
        if ($inspect.Config.Labels.psobject.Properties.Match('maintainer').Count -eq 0 -or $inspect.Config.Labels.maintainer -ne "Dynamics SMB") {
            throw "Container $imageName is not a NAV/BC container"
        }
        $navversion = "$($inspect.Config.Labels.version)-$($inspect.Config.Labels.country)"
        if ($inspect.Config.Env | Where-Object { $_ -eq "IsBcSandbox=Y" }) {
            $bcStyle = "sandbox"
			Write-Host -ForegroundColor Yellow -Object "testing4..."
        }
    }
```

```
Fetching all docker images
Fetching all docker volumes
ArtifactUrl and ImageName specified
INFO: Windows 10 21H1/21H2 images are not yet available, using 2004 as these are found to work better than 20H2 on 21H1/21H2
WARNING: Unable to find matching generic image for your host OS. Using mcr.microsoft.com/businesscentral:10.0.19041.1415
Image sandbox20.2.0.0us:sandbox-20.2.41144.41959-us-mt already exists
Using image sandbox20.2.0.0us:sandbox-20.2.41144.41959-us-mt
Creating Container vjtest1
testing4...
Style: sandbox
Multitenant: Yes
```

**Comment by vvkjndl** [comment]
@freddydk I commented Line#971 in New-NavContainer.ps1 and container style did change to onprem using sandbox image.

However, I can confirm that even with onprem style container same issue did persist for me.

```
Creating Container vjtest1
testing4...
Style: onprem
Multitenant: Yes
Version: 20.2.41144.41959
Platform: 20.0.41086.41925
Generic Tag: 1.0.2.7
Container OS Version: 10.0.19041.1415 (2004)
Host OS Version: 10.0.19044.1706 (21H2)
```

```
Windows security identifier is not supported in online environments.

Exception Script Stack Trace:
at <ScriptBlock>, <No file>: line 37

PowerShell Call Stack:
at Invoke-ScriptInBcContainer, C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.10\ContainerHandling\Invoke-ScriptInNavContainer.ps1: line 58
at Invoke-NavContainerCodeunit, C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.10\ObjectHandling\Invoke-NavContainerCodeunit.ps1: line 49
at global:Install-SetupDataPack, C:\Users\azure-admin\Desktop\SCRIPTS\SIMPMgmt.ps1: line 3952
at <ScriptBlock>, <No file>: line 1

Container Free Physical Memory: 4.8Gb

Services in container vjtest1:
- MicrosoftDynamicsNavServer$BC is Running
- MSSQL$SQLEXPRESS is Running


Relevant event log from container vjtest1:
- 20220616 11:35:55 - MicrosoftDynamicsNavServer$BC

  Server instance: BC
  Category: Security
  ClientSessionId: e7e23d47-ef17-4f96-90d1-b3eda389d40c
  ClientActivityId: 2134dd11-ca38-4de7-a3a5-60f8252c42e7
  ServerSessionUniqueId: 8dea1b61-770a-4168-b94e-3911f0798cd5
  ServerActivityId: 0a7e097c-9941-4a75-928e-7edc3e8d74ec
  EventTime: 06/16/2022 15:35:55
  Message (FaultException`1): RenameUser(String, String, String, NavUserState, DateTime, String, String, Boolean, LicenseType, Boolean, String, String, Boolean, DateTime, String, String, String, Guid, String, String, String, Nullable`1) FAILURE.
  RootException: FaultException`1
```

**Comment by vvkjndl** [comment]
@freddydk I ran the tests on a container built using onprem image. I wasn't able to reproduce the issue there.

So we can confirm that the issue tied only with container built using sandbox image.

**Comment by freddydk** [comment]
The reason is obviously that you have a built image from before you specified -runsandboxasonprem and that image is used for running the container.
If you run Flush-ContainerHelperCache - then it will remove all images not currently in use - and you should be able to re-run the script.

(or remove the -imagename parameter)

Sorry for not seeing this earlier.

**Comment by freddydk** [comment]
I have checked in the piece in ContainerHelper to allow for runSandboxAsOnPrem on New-BcImage and also transferred from New-BcContainer.
Note that if you already have an image - it will not rebuild the image just because of changing this flag, you need to change the name or tag.

**Comment by vvkjndl** [comment]
@freddydk Problem has been resolved by latest BcContainerHelper preview release.

Should I communicate any specific information to downstream developers about any possible impact of using the parameter `-runSandboxAsOnPrem`? Because when I told the developers that they can use `onprem` images to solve CodeUnit execution problems they said their testing only works on `sandbox` images.

**Comment by freddydk** [comment]
You are using sandbox demo data, so your tests should still work.
It is just running a sandbox version "as onprem", meaning that entitlements have been removed.
I think that is the only difference.
They can try it out, I think - or they can wait for a fix - I do not have an ETA on that.

**Comment by vvkjndl** [comment]
Thanks @freddydk , as always. 🥇

**Comment by vvkjndl** [comment]
@freddydk Is there any public URI where I can see the status of the bug reported to Microsoft.

**Comment by freddydk** [comment]
There is no public URI for internal bugs and no progress yet (just checked the bug)
If you need to follow along with progress and add pressure you will need to file a bug yourself.

**Comment by vvkjndl** [comment]
Please share the bug portal URI if you have it handy.

I would like to add pressure as this has an impact on us.

**Comment by vvkjndl** [comment]
@freddydk Would you be able to share the status of the bug?

I am curious to know if Microsoft declared it as `Won't Fix` or `Not a Bug`, or it is still being considered.

**Comment by vvkjndl** [comment]
@freddydk I was able to pin-point the issue to `navcontainerhelper`.

Issue is being caused by Line#85 in `Invoke-NavContainerCodeunit.ps1`.

This is how the container administrator user looks like in NAV:

```
UserSecurityID               : xxxxxxxxxxxxxxxxxxxxxxxxxxx
UserName                     : USER MANAGER\CONTAINERADMINISTRATOR
FullName                     :
State                        : Enabled
ExpiryDate                   : 1/1/0001 12:00:00 AM
WindowsSecurityID            : S-1-5-93-2-1
ChangePassword               : False
LicenseType                  : FullUser
AuthenticationEmail          :
ContactEmail                 :
ExchangeIdentifier           :
ApplicationID                : 00000000-0000-0000-0000-000000000000
Password                     :
NameIdentifier               :
AuthenticationKey            :
WebServicesKey               :
WebServicesKeyExpiryDate     : 1/1/0001 12:00:00 AM
AuthenticationObjectID       :
DirectoryPermissionSetID     :
DirectoryPermissionSetIDList : System.Byte[]
TelemetryUserID              : xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
ProfileID                    :
Company                      : xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

Notice the property, `WindowsSecurityID`. This is what's causing the command to fail.

Codeunit is executing fine at Line#83.

**Comment by vvkjndl** [comment]
Shall we stop doing the `Set-NAVServerUser` in `Invoke-NavContainerCodeunit.ps1`?

What should be our approach here?

**Comment by freddydk** [comment]
The bug is marked as resolved in NextMinor and if I run this code:

```
$credential = New-Object pscredential -ArgumentList 'admin', $secretPassword
$artifactUrl = Get-BCArtifactUrl -country de -select NextMinor -sasToken ($InsiderSasTokenSecret.SecretValue | Get-PlainText)
New-BCContainer -accept_eula `
                -Auth UserPassword `
                -artifactUrl $artifactUrl `
                -Credential $credential `
                -updateHosts
Publish-BcContainerApp -appFile "https://businesscentralapps.blob.core.windows.net/bingmaps-pte/latest/bingmaps-pte-apps.zip" -sync -install
Invoke-NavContainerCodeunit -codeunitId 74120 -CompanyName "CRONUS DE"
```

It works now (it didn't before)
Version 20.2 and 20.3 doesn't work (and it will not be backported to that) - you will have to use -runsandboxasonprem to make it work.

Version 20.4 will work again.

**Comment by freddydk** [comment]
OK
The bug is fixed in 20.4 (and 21.x)
